source/common/network/connection_impl.cc

```diff
@@ -160,7 +160,11 @@ void ConnectionImpl::close(ConnectionCloseType type) {
 }
 
 Connection::State ConnectionImpl::state() const {
-  if (fd() == -1) {
+  // TODO(sbelair2): get rid of this const_cast. It was added
+  // to demonstrate whether IoHandlePtr could be a unique_ptr rather
+  // than shared. How to get around the issue of IoHandlePtr being a
+  // non-const unique_ptr, when it is called from const methods?
```

Comment:
You should be returning a `const IoHandle&` in this case, and `fd()` should be a const method. If you do this you can remove the const_cast. Please do that in this PR.